### ISSUE TITLE
System libraries lack dwarf metadata debug info

### ISSUE THREAD

**Issue by manumartin** [emscripten-core/emscripten#9703]
Hi,

I'm having issues with sourcemaps not working while building a minimal example program with emscripten 1.38.47 + LLVM (master). To generate the sourcemaps emscripten uses the dwarf information that LLVM embeds in the object files. I think the problem is the dwarf metadata is missing stuff when targeting wasm. For example when I build this very simple program targeting both native and WASM:

main.cpp:
```
#include <iostream>

int main() {
        std::cout << "Hello world" << std::endl;
}
```

with these emcc / clang commands:
```
clang++ -o main main.cpp -Wall -Werror -O0 -frtti -g
em++ -o index.html main.cpp -g4 --profiling -s DEMANGLE_SUPPORT=1 --source-map-base "http://localhost:9000/" -Wall -Werror -O0 -frtti -g
```

I would expect some references to iostream sources inside the dwarf info and in fact I can find them in the native object but not in the wasm one:
```
root@ae212671350c:/src/emscripten-tests/cppisue# llvm-dwarfdump main | grep DW_AT_decl_file | grep iostream
                  DW_AT_decl_file	("/usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/iostream")
                DW_AT_decl_file	("/usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/iostream")
root@ae212671350c:/src/emscripten-tests/cppisue# llvm-dwarfdump index.wasm | grep DW_AT_decl_file | grep iostream
(nothing)
```

The "iostream" file exists in /src/emsdk/emsdk/upstream/emscripten/system/include/libcxx/iostream.

Since there are no references to iostream in the dwarf info the generated sourcemaps don't contain any references to iostream either. This means you can't properly debug into libcxx or see correct stacktraces in the browser. The symbols on the stacktrace are ok but not the references to the source files.

**Comment by manumartin** [comment]
For anyone interested I managed to make this work by setting the -g4 flag when building the system libraries (libstdc++). I had to edit the emscripten tools/system_libs.py script and change a cflags variable there. It's kind of a hack but I haven't found the proper way of doing this.

**Comment by kripken** [comment]
We may want to add a variation for system libraries, so they have full DWARF debug info. @sbc100 I think we discussed this in the past? I think I remember being worried about link perf if we kept debug info there all the time, but with the new variations system it should be easy.

**Comment by sbc100** [comment]
I don't think there should be much link-perf impact to add debug info to all out libraries.  The linker will ignore this info in release builds anyway.

However, if you are talking about building those libraries without optimization on that is a different matter than might warrent another build variant.    Hopefully you are just wanting debug info in the existing optimized libraries.  I think we should indeed add that..

**Comment by sbc100** [comment]
We still don't have a good answer for this ..

**Comment by kripken** [comment]
This was fixed in #13078
